fraud/bad_encoding.go

```diff
@@ -0,0 +1,219 @@
+package fraud
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+
+	format "github.com/ipfs/go-ipld-format"
+	"github.com/tendermint/tendermint/pkg/consts"
+	"github.com/tendermint/tendermint/pkg/wrapper"
+
+	"github.com/celestiaorg/rsmt2d"
+
+	pb "github.com/celestiaorg/celestia-node/fraud/pb"
+	"github.com/celestiaorg/celestia-node/ipld"
+	"github.com/celestiaorg/celestia-node/ipld/plugin"
+	"github.com/celestiaorg/celestia-node/service/header"
+)
+
+type BadEncodingProof struct {
+	BlockHeight uint64
+	// Shares contains all shares from row/col
+	// Shares that did not pass verification in rmst2d will be nil
+	// For non-nil shares MerkleProofs are computed
+	Shares []*Share
+	// Position represents the number of row/col where ErrByzantineRow/ErrByzantineColl occurred
+	Position uint8
+	isRow    bool
+}
+
+func CreateBadEncodingFraudProof(
+	ctx context.Context,
+	height uint64,
+	position uint8,
+	isRow bool,
+	eds *rsmt2d.ExtendedDataSquare,
+	roots [][]byte,
+	errShares [][]byte,
+	dag format.NodeGetter,
+) (Proof, error) {
+	shares := make([]*Share, len(errShares))
+	for index, share := range errShares {
+		if share != nil {
+			data := eds.Row(uint(index))
+			if isRow {
+				data = eds.Col(uint(index))
+			}
+
+			tree, err := buildTreeFromLeaves(ctx, data, roots[index], index, dag)
+			if err != nil {
+				return nil, err
+			}
```

Comment:
This doesn't look right. It will terminate the loop as soon as the first non-complete col/row is hit. You need to compute trees for _all_ complete col/rows. This should be a `continue`.

Reply:
resolved?